foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java

```diff
@@ -0,0 +1,318 @@
+import org.slf4j.helpers.MessageFormatter;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * FoundationDB client for YCSB framework.
+ */
+
+public class FoundationDBClient extends DB {
+  private FDB fdb;
+  private Database db;
+  private String dbName;
+  private int batchSize;
+  private int batchCount;
+  private static final String API_VERSION          = "foundationdb.apiversion";
+  private static final String API_VERSION_DEFAULT  = "520";
+  private static final String CLUSTER_FILE         = "foundationdb.clusterfile";
+  private static final String CLUSTER_FILE_DEFAULT = "./fdb.cluster";
+  private static final String DB_NAME              = "foundationdb.dbname";
+  private static final String DB_NAME_DEFAULT      = "DB";
+  private static final String DB_BATCH_SIZE_DEFAULT = "0";
+  private static final String DB_BATCH_SIZE         = "foundationdb.batchsize";
+
+  private Vector<String> batchKeys;
+  private Vector<Map<String, ByteIterator>> batchValues;
+
+  private static Logger logger = LoggerFactory.getLogger(FoundationDBClient.class);
+
+  private static final AtomicInteger INIT_COUNT = new AtomicInteger(0);
+
+  /**
+   * Initialize any state for this DB. Called once per DB instance; there is one DB instance per client thread.
+   */
+  @Override
+  public void init() throws DBException {
+    INIT_COUNT.incrementAndGet();
+    synchronized (INIT_COUNT) {
+      // initialize FoundationDB driver
+      final Properties props = getProperties();
+      String apiVersion = props.getProperty(API_VERSION, API_VERSION_DEFAULT);
+      String clusterFile = props.getProperty(CLUSTER_FILE, CLUSTER_FILE_DEFAULT);
+      String dbBatchSize = props.getProperty(DB_BATCH_SIZE, DB_BATCH_SIZE_DEFAULT);
+      dbName = props.getProperty(DB_NAME, DB_NAME_DEFAULT);
+
+      logger.info("API Version: {}", apiVersion);
+      logger.info("Cluster File: {}\n", clusterFile);
+
+      try {
+        fdb = FDB.selectAPIVersion(Integer.parseInt(apiVersion.trim()));
+        db = fdb.open(clusterFile);
+        batchSize = Integer.parseInt(dbBatchSize);
+        batchCount = 0;
+        batchKeys = new Vector<String>(batchSize+1);
+        batchValues = new Vector<Map<String, ByteIterator>>(batchSize+1);
+      } catch (FDBException e) {
+        logger.error(MessageFormatter.format("Error in database operation: {}", "init").getMessage(), e);
+        throw new DBException(e);
+      } catch (NumberFormatException e) {
+        logger.error(MessageFormatter.format("Invalid value for apiversion property: {}", apiVersion).getMessage(), e);
+        throw new DBException(e);
+      }
+    }
+  }
+
+  @Override
+  public void cleanup() throws DBException {
+    synchronized (INIT_COUNT) {
+      final int curInitCount = INIT_COUNT.decrementAndGet();
+      if (batchCount > 0) {
+        batchInsert();
+        batchCount = 0;
+      }
+      if (curInitCount == 0) {
+        try {
+          db.close();
```

Comment:
over in `init()` we did fdb.open once for each client thread, but here we're only doing a single close. is that correct?

Reply:
Thank you for your detailed comment. db.close() should be called for each client thread. I've updated the PR.